File: frontend/src/api/email.ts
import request from './request'
import type { BatchImportResult, MessageResult, Page } from './types'

export interface EmailItem {
  id: number
  email: string
  password: string
  refresh_token: string
  client_id: string
  mail_url: string
  is_used: boolean
  is_disabled: boolean
  used_at: string | null
  remark: string
  created_at: string
}

export type EmailForm = Pick<
  EmailItem,
  'email' | 'password' | 'refresh_token' | 'client_id' | 'mail_url' | 'remark'
>

export interface CheckedEmailImportItem {
  line_no: number
  email: string
  success: boolean
  message: string
  source: string
}

export interface CheckedEmailImportResult extends BatchImportResult {
  checked: number
  passed: number
  rejected: number
  checks: CheckedEmailImportItem[]
}

export interface ListEmailParams {
  page?: number
  size?: number
  keyword?: string
  remark?: string
  is_used?: boolean | null
  is_disabled?: boolean | null
}

export function listEmails(params: ListEmailParams): Promise<Page<EmailItem>> {
  return request.get('/emails', { params })
}

export function createEmail(data: EmailForm): Promise<EmailItem> {
  return request.post('/emails', data)
}

export function updateEmail(
  id: number,
  data: Partial<EmailForm> & { is_used?: boolean; is_disabled?: boolean },
): Promise<EmailItem> {
  return request.put(`/emails/${id}`, data)
}

export function deleteEmail(id: number): Promise<MessageResult> {
  return request.delete(`/emails/${id}`)
}

export function batchDeleteEmails(ids: number[]): Promise<MessageResult> {
  return request.post('/emails/batch-delete', { ids })
}

export function batchSetEmailsUsed(ids: number[], isUsed: boolean): Promise<MessageResult> {
  return request.post('/emails/batch-used', { ids, is_used: isUsed })
}

export function batchSetEmailsDisabled(
  ids: number[],
  isDisabled: boolean,
): Promise<MessageResult> {
  return request.post('/emails/batch-disabled', { ids, is_disabled: isDisabled })
}

export function batchImportEmails(
  content: string,
  onDuplicate: 'skip' | 'overwrite',
): Promise<BatchImportResult> {
  return request.post('/emails/batch-import', { content, on_duplicate: onDuplicate })
}

export function batchImportEmailsChecked(
  content: string,
  onDuplicate: 'skip' | 'overwrite',
  checkMail = true,
): Promise<CheckedEmailImportResult> {
  return request.post(
    '/emails/batch-import-checked',
    { content, on_duplicate: onDuplicate, check_mail: checkMail },
    { timeout: 600000 },
  )
}

export interface MoeMailGeneratePayload {
  api_key: string
  count: number
  domain: string
  name_prefix?: string
  expiry_time: number
  password?: string
  on_duplicate?: 'skip' | 'overwrite'
}

export function generateMoeMailEmails(data: MoeMailGeneratePayload): Promise<BatchImportResult> {
  return request.post('/emails/moemail/generate', data, { timeout: 180000 })
}

export interface MailSummary {
  id: string
  subject: string
  from_addr: string
  date: string
  folder: string
  preview: string
  is_read: boolean | null
  source: string
}

export interface MailListResult {
  success: boolean
  message: string
  source: string
  messages: MailSummary[]
}

export interface MailDetailResult {
  success: boolean
  message: string
  subject: string
  from_addr: string
  to_addr: string
  date: string
  body_html: string
  body_text: string
}

export function fetchMailMessages(emailId: number, top = 20): Promise<MailListResult> {
  return request.get(`/emails/${emailId}/messages`, { params: { top } })
}

export function fetchMailDetail(
  emailId: number,
  messageId: string,
  source: string,
): Promise<MailDetailResult> {
  return request.get(`/emails/${emailId}/message`, {
    params: { message_id: messageId, source },
  })
}

File: backend/app/schemas/email.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field


class EmailBase(BaseModel):
    email: str
    password: str = ""
    refresh_token: str = ""
    client_id: str = ""
    mail_url: str = ""
    remark: str = ""


class EmailCreate(EmailBase):
    pass


class EmailUpdate(BaseModel):
    email: str | None = None
    password: str | None = None
    refresh_token: str | None = None
    client_id: str | None = None
    mail_url: str | None = None
    is_used: bool | None = None
    is_disabled: bool | None = None
    remark: str | None = None


class CheckedEmailImportRequest(BaseModel):
    content: str
    on_duplicate: str = "skip"
    check_mail: bool = True


class CheckedEmailImportItem(BaseModel):
    line_no: int
    email: str = ""
    success: bool
    message: str
    source: str = ""


class CheckedEmailImportResult(BaseModel):
    created: int = 0
    updated: int = 0
    skipped: int = 0
    failed: int = 0
    checked: int = 0
    passed: int = 0
    rejected: int = 0
    errors: list[str] = []
    checks: list[CheckedEmailImportItem] = []


class MoeMailGenerateRequest(BaseModel):
    api_key: str = Field(min_length=1)
    count: int = Field(default=1, ge=1, le=500)
    domain: str = "edu6.site"
    name_prefix: str = ""
    expiry_time: int = 0
    password: str = ""
    on_duplicate: str = "overwrite"


class EmailOut(EmailBase):
    model_config = ConfigDict(from_attributes=True)

    id: int
    is_used: bool
    is_disabled: bool = False
    used_at: datetime | None
    created_at: datetime


class MailSummaryOut(BaseModel):
    id: str
    subject: str = ""
    from_addr: str = ""
    date: str = ""
    folder: str = ""
    preview: str = ""
    is_read: bool | None = None
    source: str = ""


class MailListOut(BaseModel):
    success: bool
    message: str
    source: str = ""
    messages: list[MailSummaryOut] = []


class MailDetailOut(BaseModel):
    success: bool
    message: str
    subject: str = ""
    from_addr: str = ""
    to_addr: str = ""
    date: str = ""
    body_html: str = ""
    body_text: str = ""
